File: leetcode-scala/src/array/ContainerWithMostWater.scala
package array

/**
 * @Project leetcode
 * @Author Ravooo
 * @Since 2020/8/31 10:40
 * @Description ContainerWithMostWater
 */
object ContainerWithMostWater {

    /**
     * 2284ms   6.67%
     * 52.7MB   10.00%
     * Violence algorithm
     * @param height
     * @return
     */
    def maxArea(height: Array[Int]): Int = {
        var res = 0
        Array.range(0, height.length).foreach(start => {
            Array.range(start + 1, height.length).foreach(end => {
                res = math.max(res, (end - start) * math.min(height(start), height(end)))
            })
        })

        res
    }

    /**
     * 584ms    100.00%
     * 51.5MB   50.00%
     * Double point
     * @param height
     * @return
     */
    def maxArea2(height: Array[Int]): Int = {
        var max = 0
        var high = 0
        var width = 0
        var start = 0
        var end = height.length - 1
        while (start < end) {
            width = end - start
            if (height(start) < height(end)) {
                high = height(start)
                start += 1
            } else {
                high = height(end)
                end -= 1
            }
            max = math.max(max, high * width)
        }

        max
    }
}

File: leetcode-java/src/classify/twopointers/ContainerWithMostWater.java
package classify.twopointers;

/**
 * @author yutiantang
 * @create 2021/3/28 4:23 PM
 */
public class ContainerWithMostWater {

    /**
     * 6ms  21.25%
     * 52MB 13.36%
     * @param height
     * @return
     */
    public int maxArea(int[] height) {
        int left = 0, right = height.length - 1, max = 0;
        while (left < right) {
            max  = Math.max(max, Math.min(height[left], height[right]) * (right - left));
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        return max;
    }

    /**
     * 优化， 减少max的计算次数
     * 2ms      99.39%
     * 51.8MB   24.03%
     * @param height
     * @return
     */
    public int maxArea2(int[] height) {
        int left = 0, right = height.length - 1, max = 0;
        while (left < right) {
            max  = Math.max(max, Math.min(height[left], height[right]) * (right - left));
            if (height[left] < height[right]) {
                int lh = height[left];
                while (left < right && height[left] <= lh) {
                    left++;
                }
            } else {
                int rh = height[right];
                while (right > left && height[right] <= rh) {
                    right--;
                }
            }
        }

        return max;
    }
}
